### PR TITLE
Features/reimplementing volunteer state

### PR DESCRIPTION
**Pull request by casaper**
# [Story in Trello]()

### Done for approval?

* [ ] Acceptance criteria are met
* [ ] Code quality checks are green
* [ ] Readme updated if needed
* [ ] Story under test
* [ ] Mobile works
* [ ] Seeds created
* [ ] Translated
* [ ] Code is reviewed

### Done after the merge?

* [ ] Deployed to staging after the merge
* [ ] Tested manually on staging

### CODE REVIEW

File: app/models/volunteer.rb [hunk 1]
```diff
@@ -28,9 +29,10 @@ class Volunteer < ApplicationRecord
 
   STATES = STATES_FOR_REVIEWED.dup.unshift(REGISTERED).freeze
 
+  enum acceptance: [:undecided, :accepted, :rejected, :resigned]
+
   belongs_to :user, -> { with_deleted }, optional: true
-  belongs_to :registrar, optional: true,
-    class_name: 'User', foreign_key: 'registrar_id'
+  belongs_to :registrar, optional: true, class_name: 'User', foreign_key: 'registrar_id'
```
Comment (toupeira) [inline app/models/volunteer.rb]: `foreign_key` is optional here, default is based on the association name (just FYI, I see you were just reformatting)

File: app/models/volunteer.rb [hunk 2]
```diff
@@ -58,18 +60,40 @@ class Volunteer < ApplicationRecord
...
-  scope :seeking_clients, (-> { where(state: SEEKING_CLIENTS) })
-
   scope :created_between, ->(start_date, end_date) { where(created_at: start_date..end_date) }
 
+  scope :with_hours, (-> { joins(:hours).distinct })
+
+  scope :all_accepted, (-> { where(acceptance: :accepted) })
+  scope :all_resigned, (-> { where(acceptance: :resigned) })
+  scope :all_rejected, (-> { where(acceptance: :rejected) })
+  scope :all_undecided, (-> { where(acceptance: :undecided) })
```
Comment (toupeira) [inline app/models/volunteer.rb]: I think the enum already adds scopes for these? (without the `all_`)

File: app/models/volunteer.rb [hunk 3]
```diff
@@ -103,22 +127,28 @@ def registered?
...
+  #   state == ACCEPTED
+  # end
+
+  # def rejected?
+  #   state == REJECTED
+  # end
+
+  # def resigned?
+  #   state == RESIGNED
+  # end
```
Comment (toupeira) [inline app/models/volunteer.rb]: Remove old code

File: app/models/volunteer.rb [hunk 4]
```diff
@@ -143,6 +173,10 @@ def to_s
 
   private
 
+  def record_acceptance_changed
+    self["#{acceptance_change[1]}_at".to_sym] = Time.zone.now if acceptance_changed?
```
Comment (toupeira) [inline app/models/volunteer.rb]: `acceptance` should do the same as `acceptance_change[1]` here and would be more obvious :)

File: app/views/volunteers/_form.html.slim [hunk 5]
```diff
@@ -1,15 +1,19 @@
...
-      = f.input :rejection_text
+- if policy(Volunteer).can_manage?
+  fieldset
+    legend= t('.acceptance.management')
+    = f.input :acceptance, collection: Volunteer.acceptance_collection, as: :radio_buttons, include_blank: false, selected: @volunteer.acceptance.to_i
+    - unless @volunteer.undecided?
+      p= "#{t_attr(:accepted_at)} #{l(@volunteer.accepted_at)}" if @volunteer.accepted_at.present?
+      p= "#{t_attr(:undecided_at)} #{l(@volunteer.undecided_at)}" if @volunteer.undecided_at.present?
+      p= "#{t_attr(:rejected_at)} #{l(@volunteer.rejected_at)}" if @volunteer.rejected_at.present?
+      p= "#{t_attr(:resigned_at)} #{l(@volunteer.resigned_at)}" if @volunteer.resigned_at.present?
```
Comment (toupeira) [inline app/views/volunteers/_form.html.slim]: Shortcut: `@volunteer.resigned_at?` etc.

File: db/migrate/20170919084512_new_volunteer_state_fields.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class NewVolunteerStateFields < ActiveRecord::Migration[5.1]
+  def change
+    change_table :volunteers do |t|
+      t.integer :acceptance, default: 0
+      t.datetime :accepted_at
+      t.datetime :rejected_at
+      t.datetime :resigned_at
+      t.datetime :undecided_at
+      t.boolean :take_more_assignments, default: false
+    end
```
Comment (toupeira) [inline db/migrate/20170919084512_new_volunteer_state_fields.rb]: 1. Do we still need the old `state` attribute?
2. Do we need to convert the old `state` into the new enum?

File: app/models/volunteer.rb (hunk 2 above)
Comment (casaper) [reply app/models/volunteer.rb]: right! 😉

File: app/models/volunteer.rb (hunk 1 above)
Comment (casaper) [reply app/models/volunteer.rb]: I just uncluttered the line, didn't change anything

File: app/models/volunteer.rb (hunk 3 above)
Comment (casaper) [reply app/models/volunteer.rb]: this is WIP!

File: app/models/volunteer.rb (hunk 4 above)
Comment (casaper) [reply app/models/volunteer.rb]: Makes sense.

File: db/migrate/20170919084512_new_volunteer_state_fields.rb (hunk 6 above)
Comment (casaper) [reply db/migrate/20170919084512_new_volunteer_state_fields.rb]: No we don't, But its WIP

File: app/models/reminder.rb [hunk 7]
```diff
@@ -11,35 +11,39 @@ def self.create_for(volunteer, assignment)
   def self.conditionally_create_reminders(time = Time.zone.now)
     logger.info("Beginning reminder run, reference Date #{time}")
     logger.flush
-
```
Comment (casaper) [inline app/models/reminder.rb]: It felt kind of a mess here... Why was Active not resolved with scope?

File: app/helpers/filter_dropdown_helper.rb [hunk 8]
```diff
@@ -97,8 +106,13 @@ def dropdown_toggler_options
...
-  def translate_value(filter_attribute, t_scope)
+  def translate_value(filter_attribute, q_filter)
     return t('all') if filter_attribute.blank?
-    t(filter_attribute, t_scope)
+    q_filter = q_filter.is_a?(Symbol) ? q_filter.to_s : q_filter.first.to_s
+    if q_filter.slice! '_true'
+      t_attr(q_filter)
+    elsif q_filter.slice! '_eq'
+      t("simple_form.options.#{controller_name.singularize}.#{q_filter}.#{filter_attribute}")
+    end
```
Comment (toupeira) [inline app/helpers/filter_dropdown_helper.rb]: See https://github.com/panter/aoz-003/pull/241#discussion_r141595300

Comment (casaper) [reply app/helpers/filter_dropdown_helper.rb]: So your refusing my PR because of this and don't review anything else in it?

Comment (toupeira) [reply app/helpers/filter_dropdown_helper.rb]: Sorry I had a meeting, didn't get around to do a full review yet :) I see it's closed now?

Comment (casaper) [reply app/helpers/filter_dropdown_helper.rb]: @toupeira sorry, i was getting frustrated about the thing. PO expected it to be finished long time ago, and as It was getting endless, seeing it going even more endless kinda gave me a ... well. whatever

Comment (casaper) [reply app/helpers/filter_dropdown_helper.rb]: But well, anyhow. I cannot do a refactoring respecting the other branch within this branch, and it should definiteley not be a acceptance criteria to this PR to have that filter thing refactored for it

File: app/models/assignment.rb [hunk 9]
```diff
@@ -52,7 +58,19 @@ class Assignment < ApplicationRecord
...
+  def ended?
+    period_end && period_end < Time.zone.now.to_date
+  end
+
+  def inactive?
+    ended? || period_start > Time.zone.now.to_date
+  end
+
+  def active?
+    ended? || period_start < Time.zone.now.to_date
```
Comment (chrysalag) [inline app/models/assignment.rb]: Maybe `!ended? && period_start < Time.zone.now.to_date`

**Comment by chrysalag** [comment]
Also if attribute_change method gives a deprecation warning like attribute_changed method, you may want to check [this](https://github.com/rails/rails/pull/25337#issuecomment-225166796),

**Pull request merged**